### ISSUE TITLE
Memory leaks in tests/test-metalink.c

### ISSUE THREAD

**Issue by rockdaboot** [rockdaboot/wget2#157]
Investigate and fix to silence the leak sanitizer.
Reproducible with
```
./configure --enable-fsanitize-asan --enable-fsanitize-ubsan
make clean
make check
```

Relevant output from tests/test-metalink.log:

```
  Testing '../../src/wget2_noinstall -d --max-threads=1 --prefer-family=ipv4 -d --tries=1 'http://localhost:33197/archiveH2.gz' 2>&1'
...
=================================================================
==12592==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 457 byte(s) in 1 object(s) allocated from:
    #0 0x7f6c4b281090 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2090)
    #1 0x7f6c4ae79147 in wget_realloc /usr/oms/src/wget2/libwget/xalloc.c:130
    #2 0x7f6c4ae64eac in wget_vasprintf /usr/oms/src/wget2/libwget/printf.c:63
    #3 0x7f6c4ae6532a in wget_aprintf /usr/oms/src/wget2/libwget/printf.c:123
    #4 0x562d0e07ed99 in main /usr/oms/src/wget2/tests/test-metalink.c:182
    #5 0x7f6c484eb2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

Direct leak of 446 byte(s) in 1 object(s) allocated from:
    #0 0x7f6c4b281090 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2090)
    #1 0x7f6c4ae79147 in wget_realloc /usr/oms/src/wget2/libwget/xalloc.c:130
    #2 0x7f6c4ae64eac in wget_vasprintf /usr/oms/src/wget2/libwget/printf.c:63
    #3 0x7f6c4ae6532a in wget_aprintf /usr/oms/src/wget2/libwget/printf.c:123
    #4 0x562d0e07ec3b in main /usr/oms/src/wget2/tests/test-metalink.c:162
    #5 0x7f6c484eb2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

Direct leak of 440 byte(s) in 1 object(s) allocated from:
    #0 0x7f6c4b281090 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2090)
    #1 0x7f6c4ae79147 in wget_realloc /usr/oms/src/wget2/libwget/xalloc.c:130
    #2 0x7f6c4ae64eac in wget_vasprintf /usr/oms/src/wget2/libwget/printf.c:63
    #3 0x7f6c4ae6532a in wget_aprintf /usr/oms/src/wget2/libwget/printf.c:123
    #4 0x562d0e07ea43 in main /usr/oms/src/wget2/tests/test-metalink.c:128
    #5 0x7f6c484eb2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

Direct leak of 317 byte(s) in 1 object(s) allocated from:
    #0 0x7f6c4b281090 in realloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2090)
    #1 0x7f6c4ae79147 in wget_realloc /usr/oms/src/wget2/libwget/xalloc.c:130
    #2 0x7f6c4ae64eac in wget_vasprintf /usr/oms/src/wget2/libwget/printf.c:63
    #3 0x7f6c4ae6532a in wget_aprintf /usr/oms/src/wget2/libwget/printf.c:123
    #4 0x562d0e07eae1 in main /usr/oms/src/wget2/tests/test-metalink.c:147
    #5 0x7f6c484eb2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

SUMMARY: AddressSanitizer: 1660 byte(s) leaked in 4 allocation(s).
```

**Comment by rootkea** [comment]
Can't test this on my system because of #148 but by looking at the code it seems we have not free'd the memory. Also maybe we should check whether `wget_aprintf` was successful or not.

**Comment by rootkea** [comment]
Edit - I switched off all the proxies, rebooted the system and now I can test this. But may be I'm doing something wrong. After configuring, make cleaning and make checking it shows all 31 tests passed. Did I miss something? How do I "activate" the sanitizer?

**Comment by rockdaboot** [comment]
You activate it as above with `./configure --enable-fsanitize-asan --enable-fsanitize-ubsan`. What are your CFLAGS (`echo $CFLAGS`) ? Here they are just `-g`. I am on AMD64, gcc 6.3.0.

**Comment by rockdaboot** [comment]
> Also maybe we should check whether wget_aprintf was successful or not.

It is always successful (error & abort on memory failure).

**Comment by rootkea** [comment]
> You activate it as above with ./configure --enable-fsanitize-asan --enable-fsanitize-ubsan

I did exactly that.

> What are your CFLAGS (`echo $CFLAGS`) ?

`echo $CFLAGS` returned nothing. But ./configure output shows CFLAGS as:

    CFlags:               -I/usr/include/p11-kit-1  -DNDEBUG    -fsanitize=undefined -fsanitize=address -fno-omit-frame-pointer  


> It is always successful (error & abort on memory failure).

But docs say:
"Prints arguments to memory and returns a pointer to allocated and 0-terminated chunk of memory or NULL on error"

**Comment by rockdaboot** [comment]
Just fixed the docs.
Maybe you can reproduce with a newer gcc !? Isn't there a gcc-6 in backports ?

**Comment by rootkea** [comment]
No. There is no candidate for `gcc` in backports.

But from looking at the `CFLAGS` and `gcc` manual I don't get this. `gcc 4.9.2` does support `-fsanitize=address`. It also supports `-fsanitize=leak` which is *not* in CFLAGS but then it also says "This option only matters for linking of executables and if neither `-fsanitize=address` nor `-fsanitize=thread` is used"

So sanitizer should work with gcc Debian stable with current `CFLAGS`.

**Comment by rootkea** [comment]
BTW why doesn't `echo $CFLAGS` print anything? I can see `CFLAGS` in `./configure --enable-fsanitize-asan --enable-fsanitize-ubsan` output.

**Comment by rockdaboot** [comment]
CFLAGS is a user variable. If you don't set it, it is empty 😉

**Comment by rockdaboot** [comment]
gcc-6 is in Debian testing...

**Comment by rootkea** [comment]
Hey @rockdaboot I don't want to sound stubborn but can I continue using Debian stable `gcc` and still sanitize `wget2` especially when I can see that the gcc 4.9.2 supports `-fsanitize=address`, `-fsanitize=leak` etc. ?

About CFLAGS I thought `echo $CFLAGS` would print `CFLAGS` value as set by `configure` script. But I was wrong. Later found out via #debian

**Comment by darnir** [comment]
@rootkea while you usually can still use your Debian stable machine, you must note that Debian Stable isn't exactly the best option to use as a development machine. Its libraries are stable, but often outdated.  And I'm not entirely sure of how to debug your issue with the sanitizers not running on `gcc 4.9.2`. See if you can manage to get it to run correctly.

the `configure` script does not change the environment in which you are working. It simply generates a `Makefile` which modifies the environment in which the compiler will be executed (the specific subshell only). Hence, you can't view it directly by running `echo $CFLAGS`. As this would require you to run configure on every new terminal and everytime you reboot the machine.

**Comment by rockdaboot** [comment]
@rootkea What Darshit says regarding the dev machine is true. We are developing for future stable releases, so we have to use the latest / recent stuff.
You can add the Debian testing repo to your sources.list and just install gcc-6 from testing and still keeping the rest at stable.
Also, you could set up a VM with a Debian testing / unstable or whatever you like for development.

**Comment by rockdaboot** [comment]
@rootkea No matter what your compiler does... did you look at the output I posted ? File(s) and line numbers have been printed. Looking at the source code immediately reveals what's happening.

**Comment by rootkea** [comment]
@rockdaboot Yes I did. And we have not free'd the memory allocated with `wget_aprintf`. I could add few `free()` calls but then since sanitize doesn't work I won't be able to confirm whether sanitizer is still crying out or not ;)

Nevertheless adding testing repo. I hope this time it works with gcc-6 :)

**Comment by rootkea** [comment]
Just reporting back. Adding testing repo was a bad idea it seems. Now `apt-get install -t testing gcc` tries to upgrade almost all of my system.

> 255 upgraded, 113 newly installed, 524 to remove and 2607 not upgraded.
Need to get 1,789 MB of archives.

I'll download the `gcc` source instead and will try to build it.

**Comment by darnir** [comment]
Or your could use a VM to compile.

**Comment by rockdaboot** [comment]
@rootkea You should have installed `apt-get install gcc-6` !!!

**Comment by rockdaboot** [comment]
`gcc` is of course deeply nested into your system, gcc-6 comes on top (hopefully). At least here I have gcc-4.4, gcc-4.5, gcc-4.6, gcc-4.7, gcc-4.8, gcc-4.9, gcc-5, gcc-6 and gcc-7 installed in parallel. gcc is just linked to gcc-6.

**Comment by rootkea** [comment]
Ok. So I built a small test program to check sanitizer options and I found that having `-fsanitize=address` option reports nothing about memory leaks (exactly as per `man gcc`). So I added `-fsanitize=leak` instead. And it does report memory leaks!

Here is one interesting bit `-fsanitize=leak` doesn't work if there's `-fsanitize=address`. Yup! Weird!

Now since our current CFLAGS has `-fsanitize=address` but not `-fsanitize=leak`. I commented `as_fn_append WARN_CFLAGS " -fsanitize=address"` in `configure` script. *And* `make check CFLAGS='-fsanitize=leak'`. And voila! Now I can see those 4 leaks from test-metalink. Yay!

**Comment by rootkea** [comment]
@rockdaboot Saw your comment just know. But now have removed `testing` repos and moreover I can see memory leaks now so probably not going that way! ;)

**Comment by rockdaboot** [comment]
@rootkea Just give it a try... if it just installs additional stuff, go ahead (if it tries to remove stuff, stop it).

**Comment by rockdaboot** [comment]
So many lessons to learn... :-)

Here comes the next: ./configure is auto-generated from ./configure.ac, so your changes will be lost after the next `autoreconf` or `bootstrap`.

See `./configure --help|grep sanitiz`. Maybe it's all there what you need !?

**Comment by rootkea** [comment]
@rockdaboot `sudo apt-get install -t testing gcc-6`

> 255 upgraded, 113 newly installed, 524 to remove and 2607 not upgraded.
Need to get 1,789 MB of archives.
After this operation, 1,637 MB disk space will be freed.
Do you want to continue? [Y/n]

Nope. I don't!

**Comment by rootkea** [comment]
I will leave that line from `configure` commented out and will do `make check CFLAGS='-fsanitize=leak` and `make check CFLAGS='-fsanitize=address` in succession.

If `configure` gets updated then "I'll find *that line*. And I will *comment* it!" :D

**Comment by rockdaboot** [comment]
Just pushed a fix for `./configure --enable-fsanitize-msan`. -fsanitize=leak was missing, instead only clang sanitizer options were checked.
@rootkea After pulling, do `./autoreconf && ./configure --enable-fsanitize-msan && make check`.
You should not need to edit ./configure any more ;-)

**Comment by rootkea** [comment]
I just added 

    free(urls[0].body);
    free(urls[2].body);
    free(urls[4].body);
    free(urls[7].body);

at the end of `test-metalink.c` but `make check CFLAGS='-fsanitize=leak'` still fails with those 4 same leaks. Am I missing something?

**Comment by rootkea** [comment]
BTW `gcc-6` looks cool. My test log doesn't show line numbers.

    ==22474==ERROR: LeakSanitizer: detected memory leaks

    Direct leak of 457 byte(s) in 1 object(s) allocated from:
    #0 0x7f9731b5c559 in __interceptor_realloc (/usr/lib/x86_64-linux-gnu/liblsan.so.0+0xb559)
    #1 0x7f97318d4483 in wget_realloc (/home/rootkea/Documents/gsoc/wget2/libwget/.libs/libwget.so.0+0x7c483)
    #2 0x7f97318c9a5f in wget_vasprintf (/home/rootkea/Documents/gsoc/wget2/libwget/.libs/libwget.so.0+0x71a5f)
    #3 0x7f97318c9c34 in wget_aprintf (/home/rootkea/Documents/gsoc/wget2/libwget/.libs/libwget.so.0+0x71c34)
    #4 0x40275f in main (/home/rootkea/Documents/gsoc/wget2/tests/test-metalink+0x40275f)
    #5 0x7f972fa46b44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)

**Comment by rockdaboot** [comment]
CFLAGS="-g" ./configure ...

**Comment by rootkea** [comment]
`test-i-https.log` shows memory leaks too. Not sure where they are from. Need to look into code.

**Comment by rootkea** [comment]
`./autoreconf` not found. Doing `./bootstrap` instead.

**Comment by rockdaboot** [comment]
sorry, autoreconf without ./

**Comment by rootkea** [comment]
Ok. So I edited `test-metalink.c` as explained above. Then I did `CFLAGS="-g"`. Then `./configure --enable-fsanitize-msan` and then `make check` but it still fails and shows those exact 4 leaks in `test-metalink.log`. Not sure why the changes in test-metalink.c are not getting reflected.

And yes. I have saved the file! ;)

**Comment by rockdaboot** [comment]
I am in the train, but had the same on my machine. Maybe the sanitizer is buggy... Check against valgrind as a second opinion.

**Comment by darnir** [comment]
> @darnir Sorry forgot to reply. I don't like using VM. They lag and constantly make you feel bad about your existing host system ;)

On a side-note, I have a feeling you're trying to use a full GUI VM from within Virtualbox or something along those lines. Try to take a look at using QEMU along with KVM. I have recently done a *lot* of development inside such environments with no performance lags whatsoever.

As @rockdaboot suggested, try running the test inside valgrind. That will give you detailed information about where the leak is originating from.

**Comment by rootkea** [comment]
Note to self: Never ever do `TESTS_ENVIRONMENT="VALGRIND_TESTS=1" make check`. Just did that and `valgrind` completely ate up my 4GB RAM + 4 GB swap space. Glad I knew about `SysReq`. I'm only going to valgrind the test in question.

**Comment by darnir** [comment]
Did you run it with valgrind when the tests were compiled with the sanitize options? Those two should never be mixed, it will break a lot of things. If not, it's a bug somewhere that should be fixed. 

On 30 March 2017 6:03:36 PM GMT+02:00, Avinash Sonawane <notifications@github.com> wrote:
>Note to self: Never ever do `TESTS_ENVIRONMENT="VALGRIND_TESTS=1" make
>check`. Just did that and `valgrind` completely ate up my 4GB RAM + 4
>GB swap space. Glad I knew about `SysReq`. I'm only going to valgrind
>the test in question.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/rockdaboot/wget2/issues/157#issuecomment-290458569

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by rootkea** [comment]
> I'll use 'TESTS_ENVIRONMENT="VALGRIND_TESTS=1" make check TESTS=test-new' as suggested by @rockdaboot on bug-wget

Nope. Didn't work. How do you guys manage to *use* valgrind. It eats all of my RAM + swap area rendering my system unusable. Had to force reboot.

@darnir I don't remember. May be that was it. I'll now do just `./configure` without any sanitize options and will pray that the valgrind the best spares my system ;)

**Comment by rootkea** [comment]
Two things
1) valgrind doesn't show those 4 memory leaks after I free'd them
2) But it still fails the test and says:  

        *** Error in `./test-metalink': double free or corruption (!prev): 0x00000000017d3c80 ***
        ../build-aux/test-driver: line 107: 20020 Aborted                 "$@" > $log_file 2>&1
        FAIL: test-metalink

Not sure how to go about this one. Will look into it later. Have to leave now.

**Comment by rootkea** [comment]
> Try to take a look at using QEMU along with KVM. I have recently done a lot of development inside such environments with no performance lags whatsoever.

Can we send github pull request from QEMU? O.o

**Comment by darnir** [comment]
Depends on how you prefer to work. :)

I see multiple different approaches here:

1. QEMU with no X support. You can push from the command line and use the web browser on your host to create a pull request.
2. QEMU with no X support. You can use (hub)[https://hub.github.com/]
3. QEMU with X support. You run a minimal graphical environment and use whatever your favourite graphical tools are. This option is generally a little slower, but never as bad as running a VM inside Virtualbox.

In the past, I've used all of the three options above for different projects. Even running Fedora with KDE inside QEMU took barely any resources on my host. Compilation was just as fast as well.

**Comment by rockdaboot** [comment]
The whole thing turns out as not so obvious, here is the solution:
- urls[x].body become overwritten when it contains {{port}}, so we need an extra variable to save the pointer for a later free().
- exit(0) in the end calls wget_test_stop_server(void) from libtest.c where urls[x].body are free'd when they have been allocated.
- to make it complete we also have to free those two header values allocated by wget_aprintf().

It's pushed now.

**Comment by darnir** [comment]
I just fixed it and got a merge conflict when trying to push my patch. Found the exact same reasons. What I don't understand is why valgrind did not catch it. It's probably a bug in valgrind.

**Comment by rockdaboot** [comment]
Yeah, the memory was definitely lost. Should have been an exit value != 0.

**Comment by darnir** [comment]
The memory was definitely lost. However, the memory leak did not occur on a codepath that was actually executed under valgrind. For the tests, valgrind is used only when executing the actual tests with the `wget2` executable. However, the leak happened entirely outside it, in the test suite scaffolding code. Since this isn't run through valgrind, there was no report about a memory leak.

The leak sanitizers on the other hand profiled the entire code and hence found this memory leak lurking in the test suite.

**Comment by rockdaboot** [comment]
They are not really *lurking*. When I wrote the tests, I didn't free certain allocations to not add irrelevant code. These are freed by exit() anyways.
And from what I saw, the Leak sanitizer just spills out 'definitely lost' chunks of memory, so that should be ok.

**Comment by rootkea** [comment]
@darnir 

> Did you run it with valgrind when the tests were compiled with the sanitize options? Those two should never be mixed, it will break a lot of things. If not, it's a bug somewhere that should be fixed.

Shouldn't `./configure` be more verbose about this? Exactly like when we use both `--enable-fsanitize-asan` and `--enable-fsanitize-msan`, `./configure` fails with this error message:

> configure: error: Address Sanitizer and Memory Sanitizer are mutually exclusive

`./configure` should also fail when we use both  `asan/msan` and `--enable-valgrind-tests`

`./configure --enable-fsanitize-msan --enable-valgrind-tests && make check` dooms the system. I didn't test but I think `./configure --enable-fsanitize-asan --enable-valgrind-tests` does the same.

We should:

1. Fail `./configure` when we see `asan/msan` and `valgrind`, spitting out the message of they being mutually exclusive instead of allowing it and then making the developer crying out blaming valgrind ;)
2. Most importantly *mention* this in `./configure --help`
I'm amazed `./configure --help` says nothing about even `asan` and `msan` being mutually exclusive. It's only when dev tries the actual command they find out this *after* the `./configure` fails!

**Comment by rootkea** [comment]
What do you all think? Should I make this happen?

**Comment by rockdaboot** [comment]
This is low priority since these options are for us developers only.
But I don't want to stop you, it is a good exercise to get into the configure.ac hassle.

> Also do --enable-fsanitize-tsan and --enable-fsanitize-ubsan share the similar relationship with --enable-valgrind-tests

I don't know, just test it :-)

**Comment by rockdaboot** [comment]
And please open a new issue for this - it is a bad habit to discuss a new issue within a closed (and/or different) one.
